kobotoolbox/kpi
# feat(organizations): add endpoints to handle organization members TASK-963

### 🗒️ Checklist

1. [x] run linter locally
2. [x] update all related docs (API, README, inline, etc.), if any
3. [x] draft PR with a title `<type>(<scope>)<!>: <title> TASK-1234`
4. [x] tag PR: at least `frontend` or `backend` unless it's global
5. [x] fill in the template below and delete template comments
6. [x] review thyself: read the diff and repro the preview as written
7. [x] open PR & confirm that CI passes
8. [ ] request reviewers, if needed
9. [ ] delete this section before merging

### 📣 Summary
- Enhance the organization member management API to support role updates and member removal via PATCH and DELETE requests, excluding member creation.


### 📖 Description
- This update introduces endpoints for managing organization members:

    - GET `/api/v2/organizations/<org-id>/members/` to list all members in an organization.
    - PATCH `/api/v2/organizations/<org-id>/members/<username>/` to update a member's role (e.g., promote to admin).
    - DELETE `/api/v2/organizations/<org-id>/members/<username>/` to remove a member from the organization.

Note: Creating members is not supported via this endpoint. Roles are restricted to 'admin' or 'member'. Including the details of invited members (those who have not yet joined the organization) is not covered in this update.


### 👷 Description for instance maintainers
- This update provides new functionality to manage organization members:

    - Role updates for members (promotion/demotion between 'admin' and 'member').
    - Member removal from an organization.
    - The endpoints are optimized to use database joins to fetch member roles efficiently without excessive database queries, ensuring minimal load.



### 👀 Preview steps
1. ℹ️ Have an account and an organization with multiple members.
2. Use the GET method to list the members of the organization.
3. Use the PATCH method to update a member's role (e.g., promote a user to admin).
4. Use the DELETE method to remove a member from the organization.
5. 🟢 Notice the endpoints behave as expected, with valid role updates and member removal.


### 💭 Notes
- The code uses database joins to optimize role determination for members, avoiding inefficient role lookups.
- Role validation restricts role assignments to 'admin' or 'member' only.
- Including invited members' details is not covered in this update. This will be implemented in a future update.
